renku/core/commands/init.py

```diff
@@ -0,0 +1,148 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Project initialization logic."""
+
+import tempfile
+from datetime import datetime, timezone
+from pathlib import Path
+
+import git
+import yaml
+
+from renku.core.management.config import RENKU_HOME
+
+TEMPLATE_MANIFEST = 'manifest.yaml'
+
+
+def fetch_template(source, ref='master', tempdir=None):
+    """Fetch the template and checkout the relevant data.
+
+    :param source: url or full path of the templates repository
+    :param ref: reference for git checkout - branch, commit or tag
+    :param tempdir: temporary work directory path
+    :return: template manifest Path
+    """
+    if tempdir is None:
+        tempdir = Path(tempfile.mkdtemp())
+
+    # clone the repo locally without checking out files
+    template_repo = git.Repo.clone_from(source, tempdir, no_checkout=True)
+
+    # fetch ref and set the HEAD
+    template_repo.remotes.origin.fetch(ref)
+    try:
+        template_repo.head.reset(template_repo.commit(ref))
+    except git.exc.BadName:
+        ref = 'origin/{0}'.format(ref)
+        template_repo.head.reset(template_repo.commit(ref))
+    git_repo = git.Git(tempdir)
+
+    # checkout the manifest
+    git_repo.checkout(TEMPLATE_MANIFEST)
+    return tempdir / TEMPLATE_MANIFEST
+
+
+def validate_template(template_path):
+    """Validate a local template.
+
+    :param template_path: path of the template to validate
+    """
+    # TODO: implement a better check
+    required_folders = [RENKU_HOME]
+    required_files = ['{0}/metadata.yml'.format(RENKU_HOME), 'Dockerfile']
+    for folder in required_folders:
+        if not Path(template_path, folder).is_dir():
+            raise ValueError(
+                'Folder {0} is required for the template to be valid'.
+                format(folder)
+            )
+    for file in required_files:
+        if not Path(template_path, file).is_file():
+            raise ValueError(
+                'File {0} is required for the template to be valid'.
+                format(file)
+            )
+    return True
+
+
+def validate_template_manifest(manifest):
+    """Validate manifet content.
+
+    :param manifest: manifest file content
+    """
+    if not isinstance(manifest, list):
+        raise ValueError((
+            'The repository doesn\'t contain a valid',
+            '"{0}" file'.format(TEMPLATE_MANIFEST)
+        ))
+    for template in manifest:
+        if not template['name']:
+            raise ValueError((
+                'Every template listed in "{0}"',
+                ' must have a name'.format(TEMPLATE_MANIFEST)
+            ))
+        for attribute in ['folder', 'description']:
+            if not template[attribute]:
```

Comment:
```suggestion
            if attribute in template:
```

Otherwise, you will get `KeyError`.

Reply:
good catch :smile: 
I guess it's `if attribute not in template:`